File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __lt__(self, other):  # Needed for heapq
        return self.val < other.val

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        minhp=[]
        for node in lists:
            if node:
                heapq.heappush(minhp,node)
        dummy=ListNode(0)
        curr=dummy
        while minhp:
            node=heapq.heappop(minhp)
            curr.next=node
            curr=curr.next
            if node.next:
                heapq.heappush(minhp,node.next)
        return dummy.next

            

        

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode curr=head;
        int ct=0;
        while(curr!=null && ct<k){
            curr=curr.next;
            ct++;
        }if(ct<k) return head;
        ListNode prev=null,next=null;
        curr=head;
        for(int i=0;i<k;i++){
            next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }
        head.next=reverseKGroup(curr,k);
        return prev;
        
    }
}
